File: examples/hfmm3d_vec_example.f
      implicit none
      integer ns,nt,nd
      double precision, allocatable :: source(:,:),targ(:,:)
      double complex, allocatable :: charge(:,:),dipvec(:,:,:)
      double complex, allocatable :: pot(:,:)
      double complex, allocatable :: pottarg(:,:)

      double precision eps
      double complex eye,zk
      integer i,j,k,idim,ier
      double precision hkrand,pi,thet,phi
      

      data eye/(0.0d0,1.0d0)/

      pi = 4.0d0*atan(1.0d0)

c
cc      initialize printing routine
c
      call prini(6,13)
      write(*,*)
      write(*,*)
      write(*,*) "================================="
      call prin2("This code is an example fortran driver*",i,0)
      call prin2("On output, the code prints sample pot,pottarg*",i,0)
      write(*,*)
      write(*,*)

      zk = 2.2d0

      ns = 10000
      nt = 10000
      
      nd = 1

      allocate(source(3,ns))
      allocate(targ(3,nt))
      allocate(charge(nd,ns))
      allocate(dipvec(nd,3,ns))
      allocate(pot(nd,ns))
      allocate(pottarg(nd,nt))


      eps = 0.51d-3

      write(*,*) "=========================================="

c
c   
c       example demonstrating use of 
c        source to source+targ, charges+dipoles, pot
c



c
cc      generate sources uniformly on the sphere 
c
c
      do i=1,ns
        thet = hkrand(0)*pi
        phi = hkrand(0)*2*pi
        source(1,i) = sin(thet)*cos(phi) 
        source(2,i) = sin(thet)*sin(phi)
        source(3,i) = cos(thet)

        do idim=1,nd
          charge(idim,i) = hkrand(0) + eye*hkrand(0)
          dipvec(idim,1,i) = hkrand(0) + eye*hkrand(0)
          dipvec(idim,2,i) = hkrand(0) + eye*hkrand(0)
          dipvec(idim,3,i) = hkrand(0) + eye*hkrand(0)
          pot(idim,i) = 0
        enddo
      enddo

      do i=1,nt
        targ(1,i) = source(1,i) + 10.0d0
        targ(2,i) = source(2,i)
        targ(3,i) = source(3,i)

        do idim=1,nd
          pottarg(idim,i) = 0
        enddo
      enddo


       call hfmm3d_st_cd_p_vec(nd,eps,zk,ns,source,charge,
     1      dipvec,pot,nt,targ,pottarg,ier)

       call prin2("potential at sources=*",pot,12)
       call prin2("potential at targets=*",pottarg,12)


      stop
      end
c----------------------------------------------------------
c
cc
c
c

File: c/hfmm3d_vec_example.c
#include "stdlib.h"
#include "stdio.h"
#include "math.h"
#include "hfmm3d_c.h"
#include "complex.h"
#include "cprini.h"

int main(int argc, char **argv)
{
  cprin_init("stdout", "fort.13");
  cprin_skipline(2);

  int ns=2000;
  int nt=1999;
  int nd=5;
  int ier=0;
  double *source = (double *)malloc(3*ns*sizeof(double));
  double *targ = (double *)malloc(3*nt*sizeof(double));

  CPX *charge = (CPX *)malloc(ns*nd*sizeof(CPX));
  CPX *dipvec = (CPX *)malloc(3*ns*nd*sizeof(CPX));

  CPX *pot = (CPX *)malloc(ns*nd*sizeof(CPX));
  CPX *pottarg = (CPX *)malloc(3*ns*nd*sizeof(CPX));

// initialize arrays
  for(int i=0;i<ns;i++)
  {
    source[3*i] = pow(rand01(),2);
    source[3*i+1] = pow(rand01(),2);
    source[3*i+2] = pow(rand01(),2);
  }
  for(int i=0;i<nd*ns;i++)
  {
    charge[i] = rand01() + I*rand01();
    dipvec[3*i] = rand01() + I*rand01();
    dipvec[3*i+1] = rand01() + I*rand01();
    dipvec[3*i+2] = rand01() + I*rand01();
  }

  for(int i=0;i<nt;i++)
  {
    targ[3*i] = rand01();
    targ[3*i+1] = rand01();
    targ[3*i+2] = rand01();
  }

  double eps = 0.5e-6;
  CPX zk = 1.1 + 0.01*I;
  cprin_message("this code is an example c driver.");
  cprin_message("on output, the code prints sample pot,pottarg");
  cprin_skipline(2);

// call the fmm routine
  hfmm3d_st_cd_p_vec_(&nd, &eps, &zk, &ns, source, charge, dipvec, 
    pot, &nt, targ, pottarg, &ier);


  cprinz("pot=",pot,12);
  cprinz("grad=",pottarg,12);

  return 0;
}
